Sqlite: add checkpoints and transactions to DBLayer

rvl opened the pull request:
Relates to issue #154.

# Overview

- Implemented saving and loading of transaction history to SQLite
- Implemented saving and loading of wallet checkpoints to SQLite, including the state for sequential scheme address discovery.

# Comments

- The SqliteSpec testing is a bit light. However, there should be enough implemented for all the DBSpec tests to work. Also I plan to finish the QSM tests tomorrow which should provide even more coverage.
- Cascading deletes are not working with persistent-sqlite, which is annoying.
- DB indexes are missing on some fields. (Needs some custom SQL, can be fixed later)

KtorZ commented on src/Cardano/Wallet/Server/Plugins.hs:
```diff
@@ -207,6 +215,44 @@ instance Buildable Servant.NoContent where
+    tls = getNodeServerTlsParams walletParams
+
+    walletToNodeArgs WalletBackendParams{..} =
+        NodeApiArgs walletNodeAddress (Just tls) False ("localhost", 4323)
```
Isn't this already somewhere in `lib` or `node` on cardano-sl :thinking: ?

KtorZ commented:
@parsonsmatt Ultimately, that's our goal yes. But for the next three releases (1.5, 1.6 and 1.7), we won't likely do that because this doesn't add any particular business value but introduces a few risks and extra complexity for Daedalus and exchanges.

We need this decoupling to start experimenting with the Rust node and the new Haskell node, but that's something we're going to do "on the side", in a "dev mode" fashion while we still release a coupled wallet (there's no way to release the wallet in a decoupled fashion ATM anyway :|)

KtorZ commented:
@Anviking So, while reviewing and testing this locally, I got mad at the CLI and thought this would just be the time to also bite the bullet and clean-up this whole thing :sweat_smile: We'll have to update docs on `cardano-sl` and probably a few nix scripts when we decide to merge this into `cardano-sl/develop` (so, for 1.6 or later). For now, we end up with something like this:

```
stack exec -- cardano-wallet-server \
  --node-id                    edge                                   \
  --system-start               0                                      \
                                                                      \
  --topology                   state-integration/topology/edge.json   \
  --configuration-file         test/integration/configuration.yaml    \
                                                                      \
  --db-path                    state-integration/db                   \
  --wallet-db-path             state-integration/wallet-db            \
  --wallet-rebuild-db                                                 \
                                                                      \
  --wallet-api-address         127.0.0.1:8090                         \
  --wallet-doc-address         127.0.0.1:8190                         \
  --node-api-address           127.0.0.1:8080                         \
  --node-doc-address           127.0.0.1:8180                         \
                                                                      \
  --tls-ca-cert                state-integration/tls/edge/ca.crt      \
  --tls-wallet-server-cert     state-integration/tls/edge/server.crt  \
  --tls-wallet-server-key      state-integration/tls/edge/server.key  \
  --tls-node-client-cert       state-integration/tls/edge/client.crt  \
  --tls-node-client-key        state-integration/tls/edge/client.key  \
```

KtorZ commented on src/Cardano/Wallet/Action.hs:
```diff
@@ -61,10 +60,8 @@ actionWithWallet
-         (BS8.unpack nodeIp, fromIntegral nodePort)
-         walletNodeTlsClientCert
-         walletNodeTlsCaCertPath
-         walletNodeTlsPrivKey
```
Small refactor here, I've favored the record alias as it makes a bit less ambiguous than 3 `FilePath` arguments.

KtorZ commented on src/Cardano/Wallet/Action.hs:
```diff
@@ -74,17 +71,16 @@ actionWithWallet
             })
         let pm = configProtocolMagic genesisConfig
         WalletLayer.Kernel.bracketPassiveWallet pm dbMode logMessage' keystore nodeState (npFInjects nodeParams) $ \walletLayer passiveWallet -> do
-            migrateLegacyDataLayer passiveWallet dbPath (getFullMigrationFlag params)
```
Same as `getWalletDbOptions`.

KtorZ commented on src/Cardano/Wallet/Action.hs:
```diff
@@ -112,13 +108,12 @@ actionWithWallet
         -- connect, but still serves the doc through TLS
         , maybe [] (pure . ("doc server",)) (Plugins.docServer params)
-        -- The monitoring API for the Core node.
-        , Plugins.monitoringServer params
```
Removed old monitoring server.

KtorZ commented on src/Cardano/Wallet/Server/CLI.hs:
```diff
@@ -51,58 +45,24 @@ data WalletDBOptions = WalletDBOptions {
--- | Converts a @GenesisKeysInclusion@ into a @Bool@.
-isDebugMode :: RunMode -> Bool
-isDebugMode ProductionMode = False
-isDebugMode DebugMode      = True
```
I've removed the debug mode. It's only usage was to disable TLS and.. well, we'll do without.

Anviking commented on src/Cardano/Wallet/Action.hs:
```diff
@@ -112,13 +108,12 @@ actionWithWallet
 
             -- Periodically compact & snapshot the acid-state database.
             , ("acid state cleanup", Plugins.acidStateSnapshots (view Kernel.Internal.wallets (snd w)) params dbMode)
+            , ("node monitoring server", Plugins.nodeAPIServer params genesisConfig ntpConfig nodeRes)
```
I called it "node monitoring server" because it was the term I'd seen everywhere, though I was a bit confused by it

@KtorZ ? I guess we (still) want to call it "monitoring server" because it is a very auxiliary feature of the node?

KtorZ replied:
Indeed @Anviking . We call it monitoring api so far because that's so-to-speak its purpose. Yet, calling that "node api" is just slightly more general.

rvl committed:
Sqlite: add checkpoints and transactions to DBLayer

akegalj commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -134,15 +163,22 @@ newDBLayer fp = do
+                Just _ -> Right <$> do
+                    -- fixme: deleteCascade is not working with persistent-sqlite.
+                    -- Therefore we need to delete related entities as well.
+                    deleteCheckpoints wid
```
it would be great if we could implement cascade delete with TH on haskell side.

My understanding is that sqlite does support delete cascade but persistent-sqlite doesn't have this implemented?

Although, I agree that first iteration should be manual and we can implement this later TM.

akegalj reviewed (commented):
I have reviewed 30% of this PR - looks good.

Have to finish it

paweljakubas reviewed (approved):
LGTM! There is weeder error (I do not get why frankly speaking) and cascade delete issue raised by Matthias and Ante that I will second - but it can definitely be addressed in another PR

rvl committed:
Sqlite: Fix use of internal constructors/accessors

- Replace Wallet with unsafeInitWallet
- Replace indexedAddresses with addresses/mkAddressPool
- Replace PendingIxs(..) with pendingIxsToList and pendingIxsFromList.
- Add Index constructor to module exports

rvl committed:
Sqlite: Remove ordering field on PendingIx table

The PendingIx set is always sorted by Index.

rvl committed:
Sqlite: Add a few column indexes

rvl committed:
Sqlite: implement withLock from DBLayer

Uses an MVar for application-level locking.

This may be possible to implement using SQLite BEGIN EXCLUSIVE
TRANSACTION but I would prefer to try and refine the DBLayer interface
first.

rvl committed:
Sqlite: replace conduit dependency with resourcet

paweljakubas committed:
Style fix to calm down hnlint
